### PR TITLE
Enhance GCS artifact repo to support passing custom client

### PR DESCRIPTION
**Pull request by smurching**
## Related Issues/PRs

Similar to #7859 , #7858 

 #xxx

## What changes are proposed in this pull request?

Adds internal-facing support for passing a custom client, for custom auth, in the `GcsArtifactRepository`

## How is this patch tested?
Updated unit tests, and verified I could manually write to UC file storage using this updated client

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 1]
```diff
@@ -18,17 +18,22 @@ class GCSArtifactRepository(ArtifactRepository):
...
+    :param artifact_uri: URI of GCS bucket
+    :param client: Optional, exposed for testing. Python module to use for GCS
+                   operations, defaults to google.cloud.storage.
+    :param gcs_client: Optional. The actual client to use for GCS operations; a default
+                       client object will be created if unspecified, using default
+                       credentials as described in https://google-cloud.readthedocs.io/en/latest/core/auth.html
     """
 
-    def __init__(self, artifact_uri, client=None):
+    def __init__(self, artifact_uri, client=None, gcs_client=None):
```
Comment (harupy) [inline mlflow/store/artifact/gcs_artifact_repo.py:29]: Can we use the existing `client` argument?

Comment (smurching) [reply mlflow/store/artifact/gcs_artifact_repo.py:29]: Good question - the existing `client` argument is exposed for tests to override/mock the GCS Python module used to instantiate a GCS client. I thought about repurposing it to allow specifying the actual storage client, but figured it wasn't worth making any breaking changes. That said, the GCS artifact repo isn't publicly documented, so I'm ok with updating to reuse `client` instead if you think it's preferable

Comment (harupy) [reply mlflow/store/artifact/gcs_artifact_repo.py:29]: @smurching Thanks for the explanation!. Let's update it. Having two arguments for specifying a GCS client is really confusing.

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 2]
```diff
@@ -60,14 +56,19 @@ def parse_gcs_uri(uri):
...
 
+        if self.client is not None:
+            return self.client
         try:
-            storage_client = self.gcs.Client()
+            return gcs_storage.Client()
         except DefaultCredentialsError:
-            storage_client = self.gcs.Client.create_anonymous_client()
-        return storage_client.bucket(bucket)
+            return gcs_storage.Client.create_anonymous_client()
```
Comment (harupy) [inline mlflow/store/artifact/gcs_artifact_repo.py:68]: Can we set `client` in the constructor?

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/7862/files#r1112665624 is addressed!

**Pull request merged**